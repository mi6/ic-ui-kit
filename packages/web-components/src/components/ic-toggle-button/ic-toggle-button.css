@import "../../global/normalize.css";

/* VARIABLES FOR IC-TOGGLE-BUTTON-GROUP */
:host {
  --button-border: var(
    --toggle-button-border,
    var(--ic-border-width) solid var(--ic-action-default)
  );
}

:host(.dark) {
  --button-border: var(
    --toggle-button-border,
    var(--ic-border-width) solid var(--ic-action-dark)
  );
}

:host(.light) {
  --button-border: var(
    --toggle-button-border,
    var(--ic-border-width) solid var(--ic-action-light)
  );
}

:host(.disabled) {
  --button-border: var(--toggle-button-border, var(--ic-border-disabled));
}

:host(.expand-toggle-group-child) ::part(button) {
  width: 100%;
  min-height: inherit;
  white-space: inherit;
  height: inherit;
}

:host ic-button {
  --icon-width: var(--ic-space-lg);
  --icon-height: var(--ic-space-lg);
}

:host(.expand-toggle-group-child) ic-button {
  min-width: 100%;
  min-height: 100%;
  white-space: normal;
  height: 100%;
}

:host(.expand-toggle-group-child) ::part(button):focus,
:host(.expand-toggle-group-child.checked) ::part(button):focus {
  box-shadow: inset 0 0 0 var(--ic-space-xxxs) var(--ic-action-default),
    var(--ic-border-focus);
}

:host(.expand-toggle-group-child.light) ::part(button):focus,
:host(.expand-toggle-group-child.dark) ::part(button):focus {
  box-shadow: inset 0 0 0 var(--ic-space-xxxs) var(--ic-action-dark),
    var(--ic-border-focus);
}

:host(:focus) {
  z-index: 1;
}

:host ::part(button) {
  border: var(--button-border);
}

:host(.checked) ::part(button) {
  background-color: var(--ic-action-default);
  color: var(--ic-architectural-white);
  box-shadow: inset 0 0 0 var(--ic-space-xxxs) var(--ic-architectural-white);
}

:host(.checked) ::part(button):focus {
  box-shadow: inset 0 0 0 var(--ic-space-xxxs) var(--ic-architectural-white),
    var(--ic-border-focus);
}

:host(.checked) ::part(button):hover {
  background-color: var(--button-default-hover);
}

:host(.checked) ::part(button):active,
:host(.icon.checked) ::part(button):active {
  background-color: var(--ic-action-default-active);
}

:host(.dark.checked) ::part(button) {
  background-color: var(--ic-action-dark);
}

:host(.dark.checked) ::part(button):hover {
  background-color: var(--ic-action-dark-hover);
}

:host(.dark.checked) ::part(button):active,
:host(.icon.dark.checked) ::part(button):active {
  background-color: var(--ic-action-dark-active);
}

:host(.light.checked) ::part(button) {
  background-color: var(--ic-action-light);
  color: var(--ic-color-primary-text);
  box-shadow: inset 0 0 0 var(--ic-space-xxxs) var(--ic-architectural-black);
}

:host(.light.checked) ::part(button):hover {
  background-color: var(--ic-action-light-hover);
}

:host(.light.checked) ::part(button):active,
:host(.icon.light.checked) ::part(button):active {
  background-color: var(--ic-action-light-active);
}

:host(.light.checked) ::part(button):focus {
  box-shadow: inset 0 0 0 var(--ic-space-xxxs) var(--ic-architectural-black),
    var(--ic-border-focus);
}

:host(.disabled) ::part(button),
:host(.icon.disabled) ::part(button) {
  border: var(--button-border);
}

:host(.disabled.checked) ::part(button),
:host(.icon.disabled.checked) ::part(button) {
  background-color: transparent;
}

:host(.disabled.checked) ::part(button),
:host(.icon.disabled.checked) ::part(button)::before {
  background-image: radial-gradient(var(--ic-architectural-400) 95%, white 20%);
  background-repeat: no-repeat;
}

:host(.icon) ::part(button) {
  border: var(--button-border);
  height: calc(var(--ic-space-xl) + var(--ic-space-xs));
  width: calc(var(--ic-space-xl) + var(--ic-space-xs));
}

:host(.icon) ::part(button):active {
  background-color: var(--ic-action-default-bg-active);
}

:host(.icon.dark) ::part(button):active {
  background-color: var(--ic-action-dark-bg-active);
}

:host(.icon.light) ::part(button):active {
  background-color: var(--ic-action-light-bg-active);
}

:host(.icon.small) ::part(button) {
  height: var(--ic-space-xl);
  width: var(--ic-space-xl);
}

:host(.icon.large) ::part(button) {
  height: var(--ic-space-xxl);
  width: var(--ic-space-xxl);
}

:host(.checked.loading) ::part(button) {
  background-color: var(--button-default-active);
}

:host(.loading) ::part(ic-loading-container) {
  --inner-color: var(--ic-architectural-white);
}

:host(.checked.loading.light) ::part(button) {
  background-color: var(--ic-action-light-active);
}

:host(.checked.loading.dark) ::part(button) {
  background-color: var(--ic-action-dark-active);
}

@media (forced-colors: active) {
  :host(.checked) ::part(button),
  :host(.checked) ::part(button):hover {
    background-color: Highlight;
  }

  :host(.checked.light) ::part(button),
  :host(.checked.light) ::part(button):hover {
    background-color: Highlight;
  }

  :host(.checked.dark) ::part(button),
  :host(.checked.dark) ::part(button):hover {
    background-color: Highlight;
  }

  :host(.checked.disabled) ::part(button) {
    background-color: GrayText;
  }
}
