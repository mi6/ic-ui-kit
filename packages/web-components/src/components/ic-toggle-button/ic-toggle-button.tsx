import {
  Component,
  Element,
  Host,
  Prop,
  Event,
  EventEmitter,
  Listen,
  h,
} from "@stencil/core";
import {
  isSlotUsed,
  onComponentRequiredPropUndefined,
  removeDisabledFalse,
} from "../../utils/helpers";
import { IcSizes, IcThemeForeground } from "../../utils/types";

/**
 * @slot icon - Content will be displayed alongside the toggle button label.
 * @slot badge - Badge component overlaying the top right of the toggle button.
 */

@Component({
  tag: "ic-toggle-button",
  styleUrl: "ic-toggle-button.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class ToggleButton {
  @Element() el: HTMLIcToggleButtonElement;

  /**
   * The accessible label that will be applied to the toggle button. This is required for the icon variant of toggle buttons.
   */
  @Prop() accessibleLabel?: string;

  /**
   * The appearance of the toggle button.
   */
  @Prop() appearance?: IcThemeForeground = "default";

  /**
   * If `true`, the toggle button will be in disabled state.
   */
  @Prop() disabled?: boolean = false;

  /**
   * If `true`, the toggle button will fill the width of the container.
   */
  @Prop() fullWidth?: boolean = false;

  /**
   * The placement of the icon in relation to the toggle button label.
   */
  @Prop() iconPlacement?: "left" | "right" | "top";

  /**
   * The label to display in the toggle button. This is required for the default variant of toggle buttons.
   */
  @Prop() label?: string;

  /**
   * If `true`, the toggle button will be in loading state.
   */
  @Prop() loading?: boolean = false;

  /**
   * The size of the toggle button to be displayed.
   */
  @Prop() size?: IcSizes = "default";

  /**
   * If `true`, the toggle button will be in a checked state.
   */
  @Prop({ mutable: true, reflect: true }) toggleChecked: boolean = false;

  /**
   * The variant of the toggle button.
   */
  @Prop({ reflect: true }) variant: "default" | "icon" = "default";

  /**
   * Emitted when the user clicks a toggle button.
   */
  @Event() icToggleChecked: EventEmitter<{
    checked: boolean;
  }>;

  componentWillLoad(): void {
    removeDisabledFalse(this.disabled, this.el);
  }

  componentDidLoad(): void {
    onComponentRequiredPropUndefined(
      [
        {
          prop: this.variant === "icon" ? this.accessibleLabel : this.label,
          propName: this.variant === "icon" ? "accessible-label" : "label",
        },
      ],
      "Toggle button"
    );
  }

  @Listen("click", { capture: true })
  handleHostClick(e: Event): void {
    if (this.disabled) {
      e.stopImmediatePropagation();
    } else if (!this.loading) {
      this.toggleChecked = !this.toggleChecked;
    }
  }

  private handleFocus = (ev: FocusEvent) => {
    ev.stopImmediatePropagation();
  };

  private handleClick = (): void => {
    !this.loading &&
      !this.disabled &&
      this.icToggleChecked.emit({
        checked: this.toggleChecked,
      });
  };

  render() {
    return (
      <Host
        class={{
          ["disabled"]: this.disabled,
          ["checked"]: this.toggleChecked,
          [`${this.appearance}`]: true,
          ["icon"]: this.variant === "icon",
          [`${this.size}`]: true,
          ["loading"]: this.loading,
        }}
        onFocus={this.handleFocus}
      >
        <ic-button
          aria-pressed={this.toggleChecked.toString()}
          variant={this.variant === "icon" ? "icon" : "secondary"}
          onClick={this.handleClick}
          title={this.accessibleLabel}
          aria-label={`${
            this.accessibleLabel ? this.accessibleLabel : this.label
          }, ${this.toggleChecked ? "ticked" : "unticked"}`}
          disabled={this.disabled}
          appearance={this.appearance}
          size={this.size}
          fullWidth={this.fullWidth}
          loading={this.loading}
          aria-disabled={`${this.disabled}`}
        >
          {this.variant !== "icon" && this.label}
          <slot />
          {isSlotUsed(this.el, "icon") && (
            <slot
              name="icon"
              slot={`${
                this.iconPlacement ? `${this.iconPlacement}-icon` : "icon"
              }`}
            ></slot>
          )}
          {isSlotUsed(this.el, "badge") && (
            <slot name="badge" slot="badge"></slot>
          )}
        </ic-button>
      </Host>
    );
  }
}
