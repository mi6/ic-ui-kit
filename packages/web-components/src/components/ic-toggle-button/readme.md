# ic-toggle-button



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute          | Description                                                                                                              | Type                              | Default     |
| ----------------- | ------------------ | ------------------------------------------------------------------------------------------------------------------------ | --------------------------------- | ----------- |
| `accessibleLabel` | `accessible-label` | The accessible label that will be applied to the toggle button. This is required for the icon variant of toggle buttons. | `string`                          | `undefined` |
| `appearance`      | `appearance`       | The appearance of the toggle button.                                                                                     | `"dark" \| "default" \| "light"`  | `"default"` |
| `checked`         | `checked`          | If `true`, the toggle button will be in a checked state.                                                                 | `boolean`                         | `false`     |
| `disabled`        | `disabled`         | If `true`, the toggle button will be in disabled state.                                                                  | `boolean`                         | `false`     |
| `fullWidth`       | `full-width`       | If `true`, the toggle button will fill the width of the container.                                                       | `boolean`                         | `false`     |
| `iconPlacement`   | `icon-placement`   | The placement of the icon in relation to the toggle button label.                                                        | `"left" \| "right" \| "top"`      | `undefined` |
| `label`           | `label`            | The label to display in the toggle button. This is required for the default variant of toggle buttons.                   | `string`                          | `undefined` |
| `loading`         | `loading`          | If `true`, the toggle button will be in loading state.                                                                   | `boolean`                         | `false`     |
| `size`            | `size`             | The size of the toggle button to be displayed.                                                                           | `"default" \| "large" \| "small"` | `"default"` |
| `variant`         | `variant`          | The variant of the toggle button.                                                                                        | `"default" \| "icon"`             | `"default"` |


## Events

| Event             | Description                                   | Type                                 |
| ----------------- | --------------------------------------------- | ------------------------------------ |
| `icToggleChecked` | Emitted when the user clicks a toggle button. | `CustomEvent<{ checked: boolean; }>` |


## Slots

| Slot      | Description                                                    |
| --------- | -------------------------------------------------------------- |
| `"badge"` | Badge component overlaying the top right of the toggle button. |
| `"icon"`  | Content will be displayed alongside the toggle button label.   |


## Dependencies

### Depends on

- [ic-button](../ic-button)

### Graph
```mermaid
graph TD;
  ic-toggle-button --> ic-button
  ic-button --> ic-loading-indicator
  ic-button --> ic-tooltip
  ic-loading-indicator --> ic-typography
  ic-tooltip --> ic-typography
  style ic-toggle-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


