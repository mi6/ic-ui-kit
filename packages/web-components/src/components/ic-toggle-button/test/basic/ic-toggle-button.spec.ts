import { newSpecPage } from "@stencil/core/testing";
import { ToggleButton } from "../../ic-toggle-button";

describe("ic-toggle-button component", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button label="Toggle"></ic-toggle-button>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render icon variant", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button variant="icon" accessible-label="Refresh the page">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            fill="#000000"
          >
            <path d="M0 0h24v24H0V0z" fill="none" />
            <path
              d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
            />
          </svg>
        </ic-toggle-button>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render slotted icon", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button label="Toggle">
          <svg
            slot="icon"
            xmlns="http://www.w3.org/2000/svg"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            fill="#000000"
          >
            <path d="M0 0h24v24H0V0z" fill="none" />
            <path
              d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
            />
          </svg>
        </ic-toggle-button>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render slotted icon with icon placement", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button label="Toggle" icon-placement="right">
          <svg
            slot="icon"
            xmlns="http://www.w3.org/2000/svg"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            fill="#000000"
          >
            <path d="M0 0h24v24H0V0z" fill="none" />
            <path
              d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
            />
          </svg>
        </ic-toggle-button>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render slotted badge", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button label="Toggle">
          <ic-badge label="1" slot="badge" variant="success"></ic-badge>
        </ic-toggle-button>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should update checked value on click", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button label="Toggle" id='test-button' onclick='alert('test')'></ic-toggle-button>`,
    });

    expect(page.rootInstance.checked).toBeFalsy();

    const element = document.getElementById("test-button");

    element.click();

    await page.waitForChanges();

    expect(page.rootInstance.checked).toBeTruthy();
  });

  it("should emit icToggleChecked on click", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button label="Toggle" id='test-button' onclick='alert('test')'></ic-toggle-button>`,
    });

    const eventSpy = jest.fn();

    page.win.addEventListener("icToggleChecked", eventSpy);

    page.rootInstance.handleClick();

    await page.waitForChanges();

    expect(eventSpy).toHaveBeenCalled();
  });

  it("should stop immediate propagation of a click event when disabled", async () => {
    const page = await newSpecPage({
      components: [ToggleButton],
      html: `<ic-toggle-button label="Toggle" id='test-button' disabled=true onclick='alert('test')'></ic-toggle-button>`,
    });

    jest.spyOn(window, "alert").mockImplementation();

    const element = document.getElementById("test-button");

    element.click();

    await page.waitForChanges();

    expect(window.alert).not.toHaveBeenCalled;
  });
});
