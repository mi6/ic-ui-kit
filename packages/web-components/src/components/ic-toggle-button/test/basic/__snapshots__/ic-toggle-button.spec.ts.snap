// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-toggle-button component should render 1`] = `
<ic-toggle-button class="default" label="Toggle" variant="default">
  <mock:shadow-root>
    <ic-button appearance="default" aria-disabled="false" aria-label="Toggle, unticked" aria-pressed="false" size="default" variant="secondary">
      Toggle
      <slot></slot>
    </ic-button>
  </mock:shadow-root>
</ic-toggle-button>
`;

exports[`ic-toggle-button component should render icon variant 1`] = `
<ic-toggle-button accessible-label="Refresh the page" class="default icon" variant="icon">
  <mock:shadow-root>
    <ic-button appearance="default" aria-disabled="false" aria-label="Refresh the page, unticked" aria-pressed="false" size="default" title="Refresh the page" variant="icon">
      <slot></slot>
    </ic-button>
  </mock:shadow-root>
  <svg fill="#000000" height="24px" viewBox="0 0 24 24" width="24px" xmlns="http://www.w3.org/2000/svg">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"></path>
  </svg>
</ic-toggle-button>
`;

exports[`ic-toggle-button component should render slotted badge 1`] = `
<ic-toggle-button class="default" label="Toggle" variant="default">
  <mock:shadow-root>
    <ic-button appearance="default" aria-disabled="false" aria-label="Toggle, unticked" aria-pressed="false" size="default" variant="secondary">
      Toggle
      <slot></slot>
      <slot name="badge" slot="badge"></slot>
    </ic-button>
  </mock:shadow-root>
  <ic-badge slot="badge" text-label="1" variant="success"></ic-badge>
</ic-toggle-button>
`;

exports[`ic-toggle-button component should render slotted icon 1`] = `
<ic-toggle-button class="default" label="Toggle" variant="default">
  <mock:shadow-root>
    <ic-button appearance="default" aria-disabled="false" aria-label="Toggle, unticked" aria-pressed="false" size="default" variant="secondary">
      Toggle
      <slot></slot>
      <slot name="icon" slot="icon"></slot>
    </ic-button>
  </mock:shadow-root>
  <svg fill="#000000" height="24px" slot="icon" viewBox="0 0 24 24" width="24px" xmlns="http://www.w3.org/2000/svg">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"></path>
  </svg>
</ic-toggle-button>
`;

exports[`ic-toggle-button component should render slotted icon with icon placement 1`] = `
<ic-toggle-button class="default" icon-placement="right" label="Toggle" variant="default">
  <mock:shadow-root>
    <ic-button appearance="default" aria-disabled="false" aria-label="Toggle, unticked" aria-pressed="false" size="default" variant="secondary">
      Toggle
      <slot></slot>
      <slot name="icon" slot="right-icon"></slot>
    </ic-button>
  </mock:shadow-root>
  <svg fill="#000000" height="24px" slot="icon" viewBox="0 0 24 24" width="24px" xmlns="http://www.w3.org/2000/svg">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"></path>
  </svg>
</ic-toggle-button>
`;
